text/3711-required-inlining.md

```diff
@@ -0,0 +1,153 @@
+- RFC PR: [rust-lang/rfcs#3711](https://github.com/rust-lang/rfcs/pull/3711)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+rustc supports inlining functions with the `#[inline]`/`#[inline(always)]`/
+`#[inline(never)]` attributes, but these are only hints to the compiler.
+Sometimes it is desirable to require inlining, not just suggest inlining:
+security APIs can depend on being inlined for their security properties, and
+performance intrinsics can work poorly if not inlined.
+
+# Motivation
+[motivation]: #motivation
+
+rustc's current inlining attributes are only hints, they do not guarantee that
+inlining is performed and users need to check manually whether inlining was
+performed. While this is desirable for the vast majority of use cases, there are
+circumstances where inlining is necessary to maintain the security properties of
+an API, or where performance of intrinsics is likely to suffer if not inlined.
+For example, consider:
+
+- Armv8.3-A's pointer authentication, when used explicitly with intrinsics[^1],
+rely on being inlined to guarantee their security properties. Rust should make
+it easier to guarantee that these intrinsics will be inlined and emit an error
+if that is not possible or has not occurred.
+- Arm's Neon and SVE performance intrinsics have work poorly if not inlined.
+Since users rely on these intrinsics for their application's performance, Rust
+should be able to warn users when these have not been inlined and performance
+will not be as expected.
+
+In general, for many intrinsics, inlining is not an optimisation but an
+important part of the semantics, regardless of optimisation level.
+
+[^1]: Armv8.3-A's pointer authentication can be used implicitly (automatic,
+hint-based) and explicitly (intrinsics). Implicit pointer authentication is
+already implemented by Rust. Explicit pointer authentication intrinsics are not
+yet available in Rust. Exposing pointer authentication intrinsics would require
+this RFC, or something providing equivalent guarantees, as a prerequisite.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are now be five variations of the `#[inline]` attribute, the
+existing bare `#[inline]`, `#[inline(always)]`, `#[inline(never)]` and
+the new  `#[inline(must)]` and `#[inline(required)]`. Bare `#[inline]`,
+`#[inline(always)]` and `#[inline(never)]` are unchanged. `#[inline(must)]`
+and `#[inline(required)]` will always attempt to inline the annotated item,
+regardless of optimization level.
+
+If it is not possible to inline the annotated item then the compiler will emit
+a lint on the caller of the annotated item that the item could not be inlined,
+providing a reason why inlining was not possible. `#[inline(must)]` emits a
+warn-by-default lint and `#[inline(required)]` emits a deny-by-default lint,
+which can be used by items which must be inlined to uphold security properties.
```

Comment:
I do not think that a deny-by-default lint is sufficiently stern for upholding security properties, especially if the point of this RFC is that auditing the assembly is not an option. A deny-by-default lint is silent when it fires in a dependency, right?

Reply:
I'm unclear if anyone outside of a crypto library would ever use this for "security". Does an everyday user need to be forcing inlining when writing a webserver or whatever the heck?

Reply:
> I do not think that a deny-by-default lint is sufficiently stern for upholding security properties, especially if the point of this RFC is that auditing the assembly is not an option. A deny-by-default lint is silent when it fires in a dependency, right?

Yeah, you need to trust your dependencies which use `inline(required)` functions to not just `#[allow(..)]`'ing those lints. I made it a lint so that you could allow/expect it yourself if limitations in the inliner were causing the lint to trigger and there was nothing you could do about that.

> I'm unclear if anyone outside of a crypto library would ever use this for "security". Does an everyday user need to be forcing inlining when writing a webserver or whatever the heck?

See https://github.com/rust-lang/rfcs/pull/3711#discussion_r1801443216 for an example of the type of function where `inline(required)` would be intended to be used. I don't anticipate that the average user would want to be using this to write their webserver, it's to support a very specific thing, which is why I'm entirely open to it being a `rustc_must_inline` attribute or being perma-unstable or something like that.

Reply:
Not trying to be too concerned about the process of it all, but would this idea better fit a Compiler MCP, rather than a general RFC? If it's intended to just eternally be some implementation detail. Generally, users expect RFCs to result in something that they can eventually use on stable. That also might reduce the bikeshed.